samples/nrf_desktop/src/modules/ble_bond.c

```diff
@@ -10,78 +10,367 @@
+		shell_print(shell, "\tIdentity %zu (%s):", i, addr_str);
+		bt_foreach_bond(i, show_single_peer, (void *)shell);
 	}
 
-	if (err) {
-		module_set_state(MODULE_STATE_ERROR);
+	return 0;
+}
+
+static int shell_remove_peers(const struct shell *shell, size_t argc,
+			      char **argv)
+{
+	return remove_peers();
+}
+
+static void cancel_operation(void)
+{
+	LOG_INF("Cancel peer operation");
+
+	state = STATE_IDLE;
+	timestamp = 0;
+
+	k_delayed_work_cancel(&timeout);
+
+	struct ble_peer_operation_event *event = new_ble_peer_operation_event();
+
+	event->op = PEER_OPERATION_CANCEL;
+	event->arg = 0;
+
+	EVENT_SUBMIT(event);
+}
+
+static u8_t next_peer_id(u8_t id)
+{
+	id++;
+
+	if (id == CONFIG_BT_MAX_PAIRED) {
+		id = 0;
+	}
+
+	return id;
+}
+
+static void select_start(void)
+{
+	LOG_INF("Start peer select");
+
+	tmp_peer_id = next_peer_id(cur_peer_id);
+
+	struct ble_peer_operation_event *event = new_ble_peer_operation_event();
+
+	event->op = PEER_OPERATION_SELECT;
+	event->arg = tmp_peer_id;
+
+	EVENT_SUBMIT(event);
+}
+
+static void select_next(void)
+{
+	LOG_INF("Cycle peers");
+
+	tmp_peer_id = next_peer_id(tmp_peer_id);
+
+	struct ble_peer_operation_event *event = new_ble_peer_operation_event();
+
+	event->op = PEER_OPERATION_SELECT;
+	event->arg = tmp_peer_id;
+
+	EVENT_SUBMIT(event);
+}
+
+static void select_confirm(void)
+{
+	LOG_INF("Select peer");
+
+	struct ble_peer_operation_event *event = new_ble_peer_operation_event();
+
+	if (tmp_peer_id != cur_peer_id) {
+		cur_peer_id = tmp_peer_id;
+
+		event->op = PEER_OPERATION_SELECTED;
+		event->arg = cur_peer_id;
+	} else {
+		event->op = PEER_OPERATION_CANCEL;
+		event->arg = 0;
+	}
+
+	EVENT_SUBMIT(event);
+}
+
+static void erase_start(void)
+{
+	LOG_INF("Start peer erase");
+
+	struct ble_peer_operation_event *event = new_ble_peer_operation_event();
+
+	event->op = PEER_OPERATION_ERASE;
+	event->arg = 0;
+
+	EVENT_SUBMIT(event);
+}
+
+static void erase_confirm(void)
+{
+	LOG_INF("Erase peer");
+
+	int err = remove_peers();
+	if (!err) {
+		struct ble_peer_operation_event *event =
+			new_ble_peer_operation_event();
+
+		event->op = PEER_OPERATION_ERASED;
+		event->arg = 0;
+
+		EVENT_SUBMIT(event);
 	} else {
 		module_set_state(MODULE_STATE_READY);
 	}
 }
 
-#if CONFIG_SHELL
-void peer_info_show(const struct bt_bond_info *info, void *user_data)
+static enum click get_click(u32_t time_diff)
 {
-	const struct shell *shell = (const struct shell *)user_data;
-	char addr[BT_ADDR_LE_STR_LEN];
+	if (time_diff < SHORT_CLICK_MAX) {
+		return CLICK_SHORT;
+	}
+	return CLICK_NONE;
+}
 
-	bt_addr_le_to_str(&info->addr, addr, sizeof(addr));
-	shell_print(shell, "\t%s", addr);
+static void handle_click(enum click click)
+{
+	if (click != CLICK_NONE) {
+		for (size_t i = 0; i < ARRAY_SIZE(state_switch); i++) {
+			if ((state_switch[i].state == state) &&
+			    (state_switch[i].click == click)) {
+				if (NULL != state_switch[i].func) {
+					state_switch[i].func();
+				}
```

Comment:
ASSERT or `else {LOG_ERR("Callback missing")}`

Reply:
My assumption was that it is perfectly fine not to have this callback. It's when you only want to change state without any effect.

Reply:
Agree, it's a state implementation using a state matrix, could be without callback.